Add more details on developer documentation (#126674)
elastic/kibana: dev_docs/contributing/documentation.mdx

title: Documentation
summary: Writing documentation during development
date: 2022-03-01
tags: ['kibana', 'onboarding', 'dev']
---

Docs should be written during development and accompany PRs when relevant. There are multiple types of documentation, and different places to add each.

## End-user documentation

Documentation about user facing features should be written in [asciidoc](http://asciidoc.org/) at [https://github.com/elastic/kibana/tree/main/docs](https://github.com/elastic/kibana//tree/master/docs)

To build the docs, you must clone the [elastic/docs](https://github.com/elastic/docs) repo as a sibling of your Kibana repo. Follow the instructions in that project’s [README](https://github.com/elastic/docs#readme) for getting the docs tooling set up.

To build the docs:

```bash
node scripts/docs.js --open
```

## REST APIs
REST APIs should be documented using the following recommended formats:

- [API doc template](https://raw.githubusercontent.com/elastic/docs/master/shared/api-ref-ex.asciidoc)
- [API object definition template](https://raw.githubusercontent.com/elastic/docs/master/shared/api-definitions-ex.asciidoc)

## Developer documentation

> Developer documentation can be segmented into two types: internal plugin details, and information on extending Kibana. Our [Kibana Developer Guide](https://docs.elastic.dev/kibana-dev-docs/getting-started/welcome) is meant to serve the latter. The guide can only be accessed internally, though the raw content is public in our [public repository]((https://github.com/elastic/kibana/tree/main/dev_docs)). We hope to eventually make this system public for the benefit of external developers as well.

Reviewer: I attempted to simplify the statement about the rendered docs being available only internally by adding the phrase "at the moment" which alludes to our intention of making this public eventually.

```suggestion
Developer documentation can be segmented into two types: internal plugin details, and information on extending Kibana. Our [Kibana Developer Guide](https://docs.elastic.dev/kibana-dev-docs/getting-started/welcome) is meant to serve the latter. The guide can only be accessed internally at the moment, though the raw content is public in our [public repository]((https://github.com/elastic/kibana/tree/main/dev_docs)).
```

When a developer first lands in our docs, think about their journey. Introduce basic concepts before diving into details. The left navigation should be set up so documents on top are higher level than documents near the bottom.

#### Think outside-in

It's easy to forget what it felt like to first write code in Kibana, but do your best to frame these docs "outside-in". Don't use esoteric, internal language unless a definition is documented and linked. The fresh eyes of a new hire can be a great asset.

## API documentation

We automatically generate <DocLink id="kibDevDocsApiWelcome" text="API documentation"/>. The following guidelines will help ensure your <DocLink id="kibPlatformIntro" section="public-plugin-api" text="public APIs" /> are useful.

If you encounter an error of the form:

<DocCallOut color="danger" title="JavaScript heap out of memory" />

You can increase [max memory](https://nodejs.org/api/cli.html#--max-old-space-sizesize-in-megabytes) for node as follows:

```bash
# As a runtime argument
node --max-old-space-size=8192 foo/bar

# As an env variable, in order to apply it systematically
export NODE_OPTIONS=--max-old-space-size=8192
```

### Code comments

> Every publicly exposed function, class, interface, type, parameter and property should have a JSDoc style comment.

Reviewer 2: 1. Perhaps we should clarify what "publicly exposed" means?
2. It might be useful to link to the ~JSDoc~ ~TSDoc~ documentation here.

```suggestion
Every function, class, interface, type, parameter and property that is exposed to other plugins should have a [TSDoc](https://tsdoc.org/)-style comment.
```

- Use `@param` tags for every function parameter.
- Use `@returns` tags for return types.
- Use `@throws` when appropriate.
- Use `@beta` or `@deprecated` when appropriate.
- Use `@removeBy {version}` on `@deprecated` APIs. The version should be the last version the API will work in. For example, `@removeBy 7.15` means the API will be removed in 7.16. This lets us avoid mid-release cycle coordination. The API can be removed as soon as the 7.15 branch is cut.

> - Use `@internal` to indicate this API item is intended for internal use only, which will also remove it from the docs.

Reviewer 2: We use `@public` in several places, it's not clear to me what the distinction is, perhaps we should mention that here too?

Author: We don't officially recommend using `@public`, it's a convention that I think only the core team uses. I think it'd be great to use, but _only_ if we had ci checks to ensure all public APIs were labeled appropriate, and fail if not. Otherwise it can't be trusted to be accurate.

Prefer types and interfaces over complex inline objects. For example, prefer:

```ts
/**
* The SearchSpec interface contains settings for creating a new SearchService, like
* username and password.
*/
export interface SearchSpec {
 /**
  * Stores the username. Duh,
  */
 username: string;
 /**
  * Stores the password. I hope it's encrypted!
  */
 password: string;
}

 /**
  * Retrieve search services
  * @param searchSpec Configuration information for initializing the search service.
  * @returns the id of the search service
  */
export getSearchService: (searchSpec: SearchSpec) => string;
```

over:

```ts
/**
  * Retrieve search services
  * @param searchSpec Configuration information for initializing the search service.
  * @returns the id of the search service
  */
export getSearchService: (searchSpec: { username: string; password: string }) => string;
```

In the former, there will be a link to the `SearchSpec` interface with documentation for the `username` and `password` properties. In the latter the object will render inline, without comments:

![prefer interfaces documentation](../assets/dev_docs_nested_object.png)

### Export every type used in a public API

> When a publicly exported API items references a private type, this results in a broken link in our docs system. The private type is, by proxy, part of your public API, and as such, should be exported.

Reviewer: Nit:

```suggestion
When a publicly exported API item references a private type, this results in a broken link in our docs system. The private type is, by proxy, part of your public API, and as such, should be exported.
```

Do:

```ts
export interface AnInterface { bar: string };
export type foo: string | AnInterface;
```

Don't:

```ts
interface AnInterface { bar: string };
export type foo: string | AnInterface;
```

### Avoid “Pick”

`Pick` not only ends up being unhelpful in our documentation system, but it's also of limited help in your IDE. For that reason, avoid `Pick` and other similarly complex types on your public API items. Using these semantics internally is fine.

![pick api documentation](../assets/api_doc_pick.png)

### Debugging tips

There are three great ways to debug issues with the API infrastructure.

1. Write a test

> [api_doc_suite.test.ts](https://github.com/elastic/kibana/blob/master/packages/kbn-docs-utils/src/api_docs/tests/api_doc_suite.test.ts) is a pretty comprehensive test suite that builds the test docs inside the [**fixtures** folder](https://github.com/elastic/kibana/tree/master/packages/kbn-docs-utils/src/api_docs/tests/__fixtures__/src).

Author: Ohh yes, great call!

Edit the code inside `__fixtures__` to replicate the bug, write a test to track what should happen, then run `yarn jest api_doc_suite`.

Once you've verified the bug is reproducible, use debug messages to narrow down the problem. This is much faster than running the entire suite to debug.

2. Use [ts-ast-viewer.com](https://ts-ast-viewer.com/#code/KYDwDg9gTgLgBASwHY2FAZgQwMbDgMQgjgG8AoOSudJAfgC44AKdIxgZximQHMBKOAF4AfHE7ckPANxkAvkA)

This nifty website will let you add some types and see how the system parses it. For example, the link above shows there is a `QuestionToken` as a sibling to the `FunctionType` which is why [this bug](https://github.com/elastic/kibana/issues/107145) reported children being lost. The API infra system didn't categorize the node as a function type node.

3. Play around with `ts-morph` in a Code Sandbox.

> You can fork [this Code Sandbox example](https://codesandbox.io/s/typescript-compiler-issue-0lkwx?file=/src/use_ts_compiler.ts) that was used to explore how various ways to create the node signature returns different results (e.g. `node.getType.getText()` shows different results than `node.getType.getText(node)`).  Here is [another messy example](https://codesandbox.io/s/admiring-field-5btxs).

Reviewer: Nit

```suggestion
You can fork [this Code Sandbox example](https://codesandbox.io/s/typescript-compiler-issue-0lkwx?file=/src/use_ts_compiler.ts) that was used to explore how various ways to create the node signature return different results (e.g. `node.getType.getText()` shows different results than `node.getType.getText(node)`).  Here is [another messy example](https://codesandbox.io/s/admiring-field-5btxs).
```

Author: Reworded a bit differently.

The code sandbox approach can be a lot faster to iterate compared to running it in Kibana.

## Example plugins

Running Kibana with `yarn start --run-examples` will include all [example plugins](https://github.com/elastic/kibana/tree/main/examples). These are tested examples of platform services in use. We strongly encourage anyone providing a platform level service or <DocLink id="kibBuildingBlocks" text="building block"/> to include a tutorial that links to a tested example plugin. This is better than relying on copied code snippets, which can quickly get out of date.

> You can also visit these [examples plugins hosted online](https://demo.kibana.dev/8.2/app/home). Note that because anonymous access is enabled, some

Reviewer: This links to a specific version (8.2) which I assume is outdated quickly (e.g. https://demo.kibana.dev/7.10/app/home does not work). Is there a possibility to keep this up-to-date automatically? (e.g. a variable here in the `.mdx` file or using a different URL structure in the nginx hosting https://demo.kibana.dev/)?

Author: Not yet, but filed https://github.com/elastic/kibana/issues/126772 so we can hopefully get this.
